### PR TITLE
[FLINK-19546][doc] Add documentation for native Kubernetes HA

### PR DESCRIPTION
**Pull request by wangyang0918**
Add documentation for FLIP-144(native Kubernetes HA).

### CODE REVIEW

**Comment by wangyang0918** [comment]
cc @tillrohrmann Could you please have a look on the documentation?

**Review by tillrohrmann — COMMENTED** [review]
Thanks for creating this PR @wangyang0918. I had a question concerning the semantics of stopping a session cluster. Please take a look.

File: docs/ops/jobmanager_high_availability.md
Comment (wangyang0918) [reply docs/ops/jobmanager_high_availability.md:272]: Hmm. After checking the implementation, the above command will only shut down the cluster and not clean up the HA state. And we are not deleting the HA related ConfigMap here.

Only when the job reached to the terminal state, the HA data will be cleaned up.

I will update the documentation.

**Comment by wangyang0918** [comment]
Address the comments and force pushed.

File: docs/ops/jobmanager_high_availability.md
Comment (sjwiesman) [inline docs/ops/jobmanager_high_availability.md:219]: I'm a little confused by this sentence, who is "we" referring to? Does this make sense? 

```suggestion
When running Flink as a Kubernetes [deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/), the `replica` count should be configured to 1.
This means that a new JobManager will be launched to take over leadership if the current one terminates exceptionally. 
```

File: docs/ops/jobmanager_high_availability.md
Comment (sjwiesman) [inline docs/ops/jobmanager_high_availability.md:229]:
```suggestion
Both session and job/application clusters support using the Kubernetes high availability service. Users just need to add the following Flink config options to [flink-configuration-configmap.yaml]({{ site.baseurl}}/ops/deployment/kubernetes.html#common-cluster-resource-definitions). All other yamls do not need to be updated.
```

File: docs/ops/jobmanager_high_availability.md
Comment (sjwiesman) [inline docs/ops/jobmanager_high_availability.md:231]: This shouldn't be Hadoop specific. 
```suggestion
<span class="label label-info">Note</span> The filesystem which corresponds to the scheme of your configured ha storage directory must be available to the runtime. Refer to [custom Flink image]({{ site.baseurl}}/ops/deployment/docker.html#customize-flink-image) and [enable plugins]({{ site.baseurl}}/ops/deployment/docker.html#using-plugins) for more information.
```

File: docs/ops/jobmanager_high_availability.md
Comment (sjwiesman) [inline docs/ops/jobmanager_high_availability.md:284]: A good rule of thumb, if you ever find yourself writing "he/she" then just use "they" instead.

Also, the word "could" implies that something only works some of the time but I don't think that's what you mean to say here. 

```suggestion
To keep HA data while restarting the Flink cluster,  simply delete the deploy (via `kubectl delete deploy <ClusterID>`). All the Flink cluster related resources will be destroyed (e.g. JobManager Deployment, TaskManager pods, services, Flink conf ConfigMap), as to not occupy the Kubernetes cluster resources. However, HA related ConfigMaps do not set the owner reference and they will be retained. When restarting the session / application, use `kubernetes-session.sh` or `flink run-application`. All the previous suspending running jobs will recover from the latest checkpoint successfully.
```

**Review by sjwiesman — COMMENTED** [review]
Cool feature, added a few comments to clarify the language.

File: docs/ops/jobmanager_high_availability.md
Comment (wangyang0918) [reply docs/ops/jobmanager_high_availability.md:219]: Hmm, I think "running Flink as a Kubernetes deployment" is not accurate. Because we are running Flink JobManager as a Kubernetes deployment. Do you think the following sentences make sense?


When running Flink JobManager as a Kubernetes [deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/), the `replica` count should be configured to 1 or greater.
* The value `1` means that a new JobManager will be launched to take over leadership if the current one terminates exceptionally. 
* The value `N`(greater than 1) means that multiple JobManagers will be launched while one is active and others are standby. Starting more than one JobManager will make the recovery faster.

**Comment by sjwiesman** [comment]
One question, can you use k8s ha with a standalone deployment on top of k8s or only with the native k8s resource manager?

**Comment by tillrohrmann** [comment]
I think it should also work with a standalone deployment. But I haven't tried it out myself. Could be a good testing task.

**Comment by sjwiesman** [comment]
Hmm, ok @tillrohrmann and @wangyang0918 I will defer to the two of you. I read the docs right now as being ambiguous as to whether or not k8s ha works with standalone clusters. I think what's here is fine to merge as is but maybe make a note to come back and update the docs if we find it doesn't work during release testing?

**Comment by wangyang0918** [comment]
@sjwiesman Thanks for raising up this question.
Before writing the documentation, I have verified that the Kubernetes HA service could work both for standalone and native 
K8s cluster. And that is by design. 

Also I have added two examples in the documentation about how to start a standalone or native cluster with HA configured. Do you think it is not clear enough?

**Comment by sjwiesman** [comment]
@wangyang0918 could you please just add a sentence that explicitly states it works for both standalone and native resource manager. It wasn’t obvious to me. After that +1 to merge

File: docs/ops/jobmanager_high_availability.md
Comment (tillrohrmann) [inline docs/ops/jobmanager_high_availability.md:286]: maybe: deploy -> deployment in the text?

Comment (tillrohrmann) [inline docs/ops/jobmanager_high_availability.md:286]: maybe: will be deleted

Comment (tillrohrmann) [reply docs/ops/jobmanager_high_availability.md:286]: For future documentation work: It is simpler to review if you put every sentence in its own line. That way it is much easier to see where the comment belongs to.

Comment (tillrohrmann) [reply docs/ops/jobmanager_high_availability.md:286]: I don't understand the subclause `, as to not occupy .... resources`.

Comment (tillrohrmann) [reply docs/ops/jobmanager_high_availability.md:286]: HA related ConfigMaps will be retained because they do not set the owner reference.

Comment (tillrohrmann) [reply docs/ops/jobmanager_high_availability.md:286]: When restarting the session / application using `kubernetes-session.sh` or `flink run-application`, all previously running jobs will be recovered and restart from the latest successful checkpoint.

**Review by tillrohrmann — CHANGES_REQUESTED** [review]
Thanks for updating this PR @wangyang0918. I had a couple of additional comments/suggestion. Please take a look.

@sjwiesman I agree that the general HA documentation and its structure is not ideal. We should rework it together with the documentation about the different deployment modes. Then the part which is introduced here should probably go to Deployment -> K8s -> HA.

**Comment by sjwiesman** [comment]
@tillrohrmann sounds good. We can discuss that as a follow up

File: docs/ops/jobmanager_high_availability.md
Comment (wangyang0918) [reply docs/ops/jobmanager_high_availability.md:286]: > I don't understand the subclause , as to not occupy .... resources.

What I want to state here is that we will free all the allocated resources from K8s so that they could be used by others. At the same time, we still have the ability to recover the Flink application at any time if you want.

**Comment by wangyang0918** [comment]
I have created a new ticket FLINK-20167 for the follow-up(rework the HA documentation).

File: docs/ops/jobmanager_high_availability.md
Comment (tillrohrmann) [reply docs/ops/jobmanager_high_availability.md:286]: I think we don't need the `as to not occupy ... resources` subclause. Are you also addressing my other comments for this paragraph?

Comment (tillrohrmann) [inline docs/ops/jobmanager_high_availability.md:286]:
```suggestion
To keep HA data while restarting the Flink cluster, simply delete the deploy (via `kubectl delete deploy <ClusterID>`). All the Flink cluster related resources will be deleted (e.g. JobManager Deployment, TaskManager pods, services, Flink conf ConfigMap). HA related ConfigMaps will be retained because they do not set the owner reference. When restarting the session / application using `kubernetes-session.sh` or `flink run-application`, all previously running jobs will be recovered and restarted from the latest successful checkpoint.
```

**Review by tillrohrmann — CHANGES_REQUESTED** [review]
I have changed my previous comments into a suggestion for change. Please take a look.

File: docs/ops/jobmanager_high_availability.md
Comment (wangyang0918) [reply docs/ops/jobmanager_high_availability.md]: Yeah. I have addressed the comments and force pushed.

**Pull request closed without merging**
